### PR TITLE
Rate limiting for frontends

### PR DESCRIPTION
**Pull request by bparli**
### Description
This is to add a rate limiting capability in Traefik, as in #643 

- Integrate the oxy ratelimit package
- Add Rate and RateLimit types for frontend configuration 
- Add rate limiting integration test
- Update glide/vendor with new package dependencies
- Add configuration example to basics doc

### CODE REVIEW

File: integration/ratelimit_test.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	c.Assert(err, checker.IsNil)
+	err = try.GetRequest("http://127.0.0.1:80/", 500*time.Millisecond, try.StatusCodeIs(http.StatusOK))
+	c.Assert(err, checker.IsNil)
+	err = try.GetRequest("http://127.0.0.1:80/", 500*time.Millisecond, try.StatusCodeIs(http.StatusTooManyRequests))
+	c.Assert(err, checker.IsNil)
+
+	time.Sleep(3 * time.Second)
+	err = try.GetRequest("http://127.0.0.1:80/", 500*time.Millisecond, try.StatusCodeIs(http.StatusOK))
+	c.Assert(err, checker.IsNil)
+}
```
Comment (nmengin) [inline integration/ratelimit_test.go]: Maybe can you set lesser values to `frontends.frontend1.ratelimit.rateset.rateset1` (average :4 and burst 5?) in the toml file and check if this rateset is respected too in the `TestSimpleConfiguration` test or in another one?

Comment (bparli) [reply integration/ratelimit_test.go]: 👍 You're right.  Its a good idea to test that first time period too.

File: glide.lock [hunk 2]
```diff
@@ -331,7 +337,7 @@ imports:
 - name: github.com/Masterminds/semver
   version: 59c29afe1a994eacb71c833025ca7acf874bb1da
 - name: github.com/Masterminds/sprig
-  version: 9526be0327b26ad31aa70296a7b10704883976d5
+  version: e039e20e500c2c025d9145be375e27cf42a94174
```
[the remark this answers is not in the document]
Comment (bparli) [reply glide.lock]: I agree this update doesn't seem to be necessary.  After looking at this again, I'm actually sure not how to handle it though.  For Masterminds/sprig, the versions don't seem to match between glide.yml and glide.lock.  I'm not super familiar with glide, but I guess thats why glide went ahead with the update? Any advice is appreciated

**Comment by bparli** [comment]
Thanks for the review and feedback guys.  I think I have glide straightened out now

File: server/server.go [hunk 3]
```diff
@@ -890,6 +891,30 @@ func (server *Server) loadConfig(configurations types.Configurations, globalConf
 						}
 					}
 
+					if frontend.RateLimit != nil && len(frontend.RateLimit.RateSet) > 0 {
```
[the remark this answers is not in the document]
Comment (bparli) [reply server/server.go]: I think I have this comment addressed with the server.buildRateLimiter function I added.  Of course, let me know if you had something else in mind though.

[the remark this answers is not in the document]
Comment (ldez) [reply server/server.go]: you did exactly what I wanted👍

**Comment by trondhindenes** [comment]
I can't find the doc changes in this PR at https://docs.traefik.io/basics/, is there any reason for that?

**Comment by ldez** [comment]
@trondhindenes because it's a 1.5 features.
If you want to try 1.5, you can find the documentation here: https://docs.traefik.io/archive/ or http://v1-5.archive.docs.traefik.io/

**Comment by trondhindenes** [comment]
gotcha. Thanks!

**Comment by tobernguyen** [comment]
How can I use this in Traefik Ingress Controller on my Kubernetes cluster? Is there any way to set the rate limit using Annotations?

**Comment by ldez** [comment]
@tobernguyen For now, you cannot define rate limit with annotations.
This feature (k8s annotations for rate limit) come in v1.6: #2612

**Comment by tobernguyen** [comment]
@ldez that's awesome! Thank you very much.

**Pull request merged**